[FXML-1991] Quantize and Dequantize Definitions

cmcgirr-amd opened the pull request:
Introducing the Quantize and Dequantize operations to the XTenNN dialect.

The idea is to be able to represent signed and unsigned per tensor quantization with any bitwidth. Note we currently model the zero point of the linear quantization as always zero. In the future this can be extended. 

We also restrict the scale to be a power of two value for the moment. 

We will use the `si<>` and `ui<>` types to explicitly denote signed and unsigned. Rather than some dialects leveraging the sign-less `i<>` type as a signed integer. 

Conversion to these operations from other dialects will have to take that into account.

The idea will also be to fold `Clip/Clamp` into a `xten_nn.quantize` operator using the bitwidth of the output.

Commit:
feat(Quantization): add operation definitions

Add definitions for quantize and dequantize operations that can
convert float tensors from and to integer tensors.

Note, the types supported are tied highly to the ONNX representation
and limited to zero_point == 0 and scale in power of two values.

Signless integers are not supported as quantization specifies the
signedness of the tensor.

Commit:
test(Quantization): add operation tests

Ensure that the defined operations behave as expected. E.g. type and
shape restrictions are applied correctly.

Commit:
test(Quantization): add unittests for power-of-two check

Commit:
test(Quantization): add more lit tests for different bitwidths

flemairen6 commented on include/xten/Dialect/XTenNN/IR/XTenNNTypes.h:
```diff
@@ -0,0 +1,27 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2022 Advanced Micro Devices, Inc.
```
```suggestion
// (c) Copyright 2023 Advanced Micro Devices, Inc.
```

flemairen6 commented on include/xten/Dialect/XTenNN/IR/XTenNNTypes.h:
```diff
@@ -0,0 +1,27 @@
+///\param attr the floating point attribute
+///\return true when the floating point is a power-of-two
+///\return false otherwise or when the attribute is not a FloatAttr
+bool isPowerOfTwoFloat(mlir::Attribute attr);
```
Do we need to actively make sure that we get power of two as input, or is it just VAIQ that only gives us power of 2 for the moment?

cmcgirr-amd replied:
For now we will restrict ourselves to only power-of-two because that is what we can support at the moment. And this can give us a guarantee that whenever we see this operation its always a power-of-two.

VAIQ can output other scale factors I believe, but for now we restrict ourselves to keep it simple.

Commit:
test(Unittests): add cmake target and update check-all

Commit:
refactor(Unittests): use mlir directory structure

Commit:
refactor(CMakeLists.txt): use the LIT umbrella function

to collect and group tests into one target

Commit:
fix(XTenNNTypes.h): correct LICENSE year

maxbartel commented on include/xten/Dialect/XTenNN/IR/XTenNNOps.td:
```diff
@@ -96,4 +97,56 @@ def XTenNN_OutputOp : XTenNN_Op<"output", [
+
+  let arguments = (ins 
+    F32Tensor:$input,
+    XTeNN_F32AttrPowerOfTwo:$scale
```
I am not sure if we should still have an input called `zero_point`,but verify if to be 0. It is a bit confusing without it IMO

maxbartel commented on include/xten/Dialect/XTenNN/IR/XTenNNOps.td:
```diff
@@ -96,4 +97,56 @@ def XTenNN_OutputOp : XTenNN_Op<"output", [
+
+  let arguments = (ins 
+    XTenNN_AnyTensorSignedOrUnsignedInteger:$input,
+    XTeNN_F32AttrPowerOfTwo:$scale
```
Why do we save the scale as an F32?  And why not just the exponent as an integer? The check would then be at conversion time. (The rescale op in TOSA also only saves the exponent)

cmcgirr-amd replied:
The idea was to keep it a little general and open. 

I believe we can't simply take the exponent we also need cases when the mantissa has a single bit set. E.g. smallest positive subnormal number. (https://en.wikipedia.org/wiki/Single-precision_floating-point_format#Notable_single-precision_cases)

maxbartel commented on lib/Dialect/XTenNN/IR/XTenNNTypes.cpp:
```diff
@@ -0,0 +1,46 @@
+
+namespace amd::xten_nn {
+
+bool isPowerOfTwoFloat(Attribute attr) {
+  // Only handle floating point attributes
+  if (!attr.isa<FloatAttr>()) {
+    return false;
+  }
+  auto floatValue = attr.cast<FloatAttr>().getValue();
+
+  // Negative values, non finite and not IEEE cannot be handled.
+  if (!floatValue.isFinite() || !floatValue.isIEEE() ||
+      floatValue.isNegative()) {
+    return false;
+  }
+
+  // Here we convert the float value to its components with the following
+  // equation:
+  //     floatValue = normalized_mantissa * 2^exponent;
+  int exponent = 0;
+  APFloat mantissa =
+      frexp(floatValue, exponent, APFloat::roundingMode::TowardZero);
+
+  // Check that the mantissa is zero. If the normalized mantissa value is 0.5
+  // then the mantissa was zero.
+  return mantissa.convertToFloat() == 0.5;
+}
```
Again same question, if negative values are not allowed, why do we not save just the exponent as an integer? All this runtime checking can then be simplified to a check that it is indeed an int.

maxbartel commented on test/Dialect/XTenNN/dequantize.mlir:
```diff
@@ -0,0 +1,67 @@
+// RUN: aten-opt %s -split-input-file -verify-diagnostics
+
+func.func @valid_dequantize_op_signed(%arg0: tensor<1x2xsi8>) -> tensor<1x2xf32> {
```
How to you plan to make this compatible with TOSA in the long run? AFAIK the work to move TOSA to signed types was mentioned but is not planned in the near future.

cmcgirr-amd replied:
Because we will always cast back to float our network tosa ops will be work. The only issue I see is the use of unsigned in the decomposition of these ops. This we need to somehow fix. 

For now when converting the signed case we have to assume `i<> == si<>` ONNX has the same problem.

mgehre-amd commented on include/xten/Dialect/XTenNN/IR/XTenNNOps.td:
```diff
@@ -96,4 +97,56 @@ def XTenNN_OutputOp : XTenNN_Op<"output", [
+      y = round( x * scale ) + zero_point
+
+      Where scale is an attribute of the operation in f32 and is a power-of-two value.
+      While the zero_point is always assumed to be zero.
```
It's a little odd to have a zero_point in the formula but then says it's not there

cmcgirr-amd replied:
That is true, but I wanted to keep the operation as simple as possible while still documenting where it is coming from. I think I will add a zero_point but constrain the value.

mgehre-amd commented on include/xten/Dialect/XTenNN/IR/XTenNNOps.td:
```diff
@@ -96,4 +97,56 @@ def XTenNN_OutputOp : XTenNN_Op<"output", [
+  }];
+
+  let arguments = (ins 
+    F32Tensor:$input,
```
What about F64, F16, BFloat16 inputs? Is there a reason not to allow any floating point input?

cmcgirr-amd replied:
The reason is this is not within scope. For the foreseeable future we will deal with FP32->INT8/UINT8

mgehre-amd commented on include/xten/Dialect/XTenNN/IR/XTenNNOps.td:
```diff
@@ -96,4 +97,56 @@ def XTenNN_OutputOp : XTenNN_Op<"output", [
+
+  let arguments = (ins 
+    F32Tensor:$input,
+    XTeNN_F32AttrPowerOfTwo:$scale
```
If this must be a power of two, then it's also going to be an integer value, right? Why not represent it as integer attribute then?

cmcgirr-amd replied:
We can have it as an integer, that is not a problem.

mgehre-amd commented on unittests/Dialect/XTenNN/IR/TestXTenNNTypes.cpp:
```diff
@@ -0,0 +1,77 @@
+auto getPowerOfTwoTestCases() {
+  mlir::MLIRContext context;
+  return testing::Values<TestValue>(
+      TestValue({1.0, true}), TestValue({0.125, true}),
```
Oh, you are not actually checking that the scale is a power of two, but a reciprocal of a power of two.
Can we rename the functions and comments to say that?

mgehre-amd replied:
Alternatively, let the ops take the `inv_scale` as attribute (the inverse of the scale), which can then be an integer type and actually a power of two.

cmcgirr-amd replied:
No we check both the variations. I have a test case for `16.0`.

mgehre-amd commented:
My general question is whether the "power of two" constraint is going to be there in the long term. If not, I'd not put them in the operator definition, and instead check that those are powers of two somewhere else; e.g. in the pass where we introduce those ops into IR.

- [x] question

cmcgirr-amd commented:
> My general question is whether the "power of two" constraint is going to be there in the long term. If not, I'd not put them in the operator definition, and instead check that those are powers of two somewhere else; e.g. in the pass where we introduce those ops into IR.
> 
> * [x]  question

For the foreseeable future, we will only deal in power-of-two values for scaling. There are no plans to change this.

cmcgirr-amd commented:
I think we should discuss this next week a bit more in detail and decide what we what to represent with these operations.

Commit:
refactor(Quantization): redefine ops using integer scale

where the scale now represents a shift and the equation for quantization simplies to round(x*2^shift)

Commit:
feat(XTenNNOps.td): update description and interface of QDQ ops

refined the description of the operator and implemented more interface constraints.

gargaroff commented on include/xten/Dialect/XTenNN/IR/XTenNNOps.td:
```diff
@@ -96,4 +97,55 @@ def XTenNN_OutputOp : XTenNN_Op<"output", [
+  let description = [{
+    Quantizes a given float32 tensor into a signed or unsigned integer tensor of given width.
+
+    Applies the following to linear quantization the input tensor x:
```
```suggestion
    Applies the following linear quantization to the input tensor x:
```

gargaroff commented on include/xten/Dialect/XTenNN/IR/XTenNNOps.td:
```diff
@@ -96,4 +97,55 @@ def XTenNN_OutputOp : XTenNN_Op<"output", [
+  let description = [{
+    Dequantizes a signed/unsigned integer tensor of given bitwidth to a float32 tensor.
+
+    Applies the following to linear dequantization the input tensor x:
```
```suggestion
    Applies the following linear dequantization to the input tensor x:
```

gargaroff commented on include/xten/Dialect/XTenNN/IR/XTenNNOps.td:
```diff
@@ -96,4 +97,55 @@ def XTenNN_OutputOp : XTenNN_Op<"output", [
+  let assemblyFormat = [{ `(`$input `:` type($input)`)` attr-dict `->` type($output) }];
+}
+
 #endif // XTENNN_OPS
```
File is not ending in an empty line

gargaroff commented on include/xten/Dialect/XTenNN/IR/XTenNNTypes.td:
```diff
@@ -0,0 +1,21 @@
+    [], 
+    "signed-or-unsigned-tensor">;
+
+#endif // XTENNN_TYPES
```
File is missing an empty line at the end

gargaroff commented on lib/Dialect/XTenNN/IR/CMakeLists.txt:
```diff
@@ -3,7 +3,7 @@
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 #
-# (c) Copyright 2022 Advanced Micro Devices, Inc.
+# (c) Copyright 2022-2023 Advanced Micro Devices, Inc.
```
Are we supposed to update this copyright header?

cmcgirr-amd replied:
Maybe not since the refactor.

gargaroff commented on include/xten/Dialect/XTenNN/IR/XTenNNTypes.td:
```diff
@@ -0,0 +1,21 @@
+//===-- XTenNNTypes.td - XTenNN Type definitions *-------- tablegen -*------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2023 Advanced Micro Devices, Inc.
```
Are we supposed to add a copyright header in this repo?

cmcgirr-amd replied:
Yes we are as this is a public repo.

gargaroff commented on test/CMakeLists.txt:
```diff
@@ -4,6 +4,7 @@
 # (c) Copyright 2021 Xilinx Inc.
+# (c) Copyright 2023 Advanced Micro Devices, Inc.
 
 configure_lit_site_cfg(
```
Can't add a comment to the line above for some reason. But same question regarding the copyright

gargaroff commented on test/Dialect/XTenNN/dequantize.mlir:
```diff
@@ -0,0 +1,67 @@
+func.func @sixteen_bitwidth(%arg0: tensor<1x2xsi16>) -> tensor<1x2xf32> {
+    %result = xten_nn.dequantize (%arg0: tensor<1x2xsi16>) {shift = -1: si32} -> tensor<1x2xf32>
+    return %result : tensor<1x2xf32>
+}
```
Missing empty line

gargaroff commented on test/Dialect/XTenNN/quantize.mlir:
```diff
@@ -0,0 +1,65 @@
+func.func @sixteen_bitwidth(%arg0: tensor<1x2xf32>) -> tensor<1x2xsi16> {
+    %result = xten_nn.quantize (%arg0: tensor<1x2xf32>) {shift = -1: si32} -> tensor<1x2xsi16>
+    return %result : tensor<1x2xsi16>
+}
```
Missing empty-line

gargaroff reviewed (approved):
Minor nits, but overall looks good.

cmcgirr-amd committed:
Update include/xten/Dialect/XTenNN/IR/XTenNNOps.td

Co-authored-by: Dominik Montada <dominik.montada@amd.com>

cmcgirr-amd committed:
Update include/xten/Dialect/XTenNN/IR/XTenNNOps.td

Co-authored-by: Dominik Montada <dominik.montada@amd.com>

Commit:
fix(CMakeLists.txt): correct copy right as file was not touched.

cmcgirr-amd commented:
For the file ending with a new line comments, I would rather have automation fix it in another PR than correct these files here. All the files seem to not adhere to this style in this repo.

Commit:
fix(CMakeLists.txt): add missing copyright year

Commit:
docs(XTenNNOps.td): update copyright year
